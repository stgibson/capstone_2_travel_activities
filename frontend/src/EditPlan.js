import React, { useState, useEffect } from "react";
import { useParams } from "react-router-dom";
import { useSelector } from "react-redux";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Day from "./Day";

/**
 * Component for displaying and editing a travel plan
 * @returns JSX code for rendering page of a travel plan
 */
const EditPlan = () => {
  // days should be arr of objects keys number and activities
  const [days, setDays] = useState([]);
  const { id } = useParams();
  const plan = useSelector(store => store.plans[id]);

  // sort days by number, and make activities prop arr
  useEffect(() => {
    const daysFromPlan = [];
    // numbers should be 1 to # of days
    if (plan && plan.days) {
      for (let i = 1; i <= Object.keys(plan.days).length; i++) {
        const day = plan.days[i];
        const activities = Object.keys(day).map(activityId => day[activityId]);
        daysFromPlan.push({ number: i, activities });
      }
      setDays(daysFromPlan);
    }
  }, [plan]);

  return (
    <div>
      <h2>{ plan ? plan.name : "Loading..." }</h2>
      <Row>
        {
          days.map(day => (
            <Col key={ day.number } xs={ 3 }>
              <Day
                day={ day }
                planId={ plan.id }
              />
            </Col>
          ))
        }
      </Row>
    </div>
  );
};

export default EditPlan;