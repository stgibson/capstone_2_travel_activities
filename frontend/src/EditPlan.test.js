import React from "react";
import { MemoryRouter, Route } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import EditPlan from "./EditPlan";

const activities = [
  { id: 1, name: "Test Activity 1" },
  { id: 2, name: "Test Activity 2" },
  { id: 3, name: "Test Activity 3" }
];
const plans = {
  1: {
    id: 1,
    name: "Test Plan 1",
    days: { 1: { 1: activities[0], 2: activities[1] }, 2: {} }
  },
  2: { id: 2, name: "Test Plan 2", days: { 1: {}, 2: {} } }
};

const mockStore = configureMockStore();
const store = mockStore({ activities, plans });

it("renders without crashing", () => {
  render(
    <Provider store={ store }>
      <MemoryRouter initialEntries={ ["/plans/1"] }>
        <Route path="/plans/:id"><EditPlan /></Route>
      </MemoryRouter>
    </Provider>
  );
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <Provider store={ store }>
      <MemoryRouter initialEntries={ ["/plans/1"] }>
        <Route path="/plans/:id"><EditPlan /></Route>
      </MemoryRouter>
    </Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
