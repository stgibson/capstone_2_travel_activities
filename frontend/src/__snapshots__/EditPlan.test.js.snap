// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Test Plan 1
    </h2>
    <div
      class="row"
    >
      <div
        class="col-md-3 col-sm-4 col-6"
      >
        <div
          class="Day"
          id="1"
        >
          <h3>
            1
          </h3>
          <div>
            <div
              class="Activity"
              id="1"
            >
              <a
                class="Activity-link"
                href="/activities/1"
              >
                Test Activity 1
              </a>
              <button
                class="btn btn-warning btn-sm"
                type="button"
              >
                Remove
              </button>
            </div>
          </div>
          <div>
            <div
              class="Activity"
              id="2"
            >
              <a
                class="Activity-link"
                href="/activities/2"
              >
                Test Activity 2
              </a>
              <button
                class="btn btn-warning btn-sm"
                type="button"
              >
                Remove
              </button>
            </div>
          </div>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="activity"
              />
              <select
                class="form-control"
                id="activity"
              >
                <option
                  value="1"
                >
                  Test Activity 1
                </option>
                <option
                  value="2"
                >
                  Test Activity 2
                </option>
                <option
                  value="3"
                >
                  Test Activity 3
                </option>
              </select>
            </div>
          </form>
          <button
            class="btn btn-primary"
            type="button"
          >
            Add Activity
          </button>
        </div>
      </div>
      <div
        class="col-md-3 col-sm-4 col-6"
      >
        <div
          class="Day"
          id="2"
        >
          <h3>
            2
          </h3>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="activity"
              />
              <select
                class="form-control"
                id="activity"
              >
                <option
                  value="1"
                >
                  Test Activity 1
                </option>
                <option
                  value="2"
                >
                  Test Activity 2
                </option>
                <option
                  value="3"
                >
                  Test Activity 3
                </option>
              </select>
            </div>
          </form>
          <button
            class="btn btn-primary"
            type="button"
          >
            Add Activity
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
